### PR TITLE
Netezza Functions added - PI, COT, INT2SHR, INT2SHL, OCTET_LENGTH, IN…

### PR DESCRIPTION
**Pull request by rohitgaurDatametica**
…T8XOR, BITWISE_XOR, BITWISE_OR, BITWISE_AND

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 1]
```diff
@@ -864,6 +865,30 @@ private void unparseOtherFunction(SqlWriter writer, SqlCall call, int leftPrec,
...
       unparseRegexMatchCount(writer, call, leftPrec, rightPrec);
       break;
+    case "COT":
+      unparseCot(writer, call, leftPrec, rightPrec);
+      break;
+    case "BITWISE_AND":
+      unparseBitwiseAnd(writer, call.operand(0), call.operand(1), leftPrec, rightPrec);
+      break;
+    case "BITWISE_OR":
+      unparseBiwiseFunctions(writer, call, "|", leftPrec, rightPrec);
```
Comment (vaibhavjain-dm) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:875]: Spelling mistake.

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 2]
```diff
@@ -864,6 +865,30 @@ private void unparseOtherFunction(SqlWriter writer, SqlCall call, int leftPrec,
     case "REGEXP_MATCH_COUNT":
       unparseRegexMatchCount(writer, call, leftPrec, rightPrec);
       break;
+    case "COT":
+      unparseCot(writer, call, leftPrec, rightPrec);
+      break;
+    case "BITWISE_AND":
+      unparseBitwiseAnd(writer, call.operand(0), call.operand(1), leftPrec, rightPrec);
```
Comment (vaibhavjain-dm) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:872]: Why passing call.operand(0) and call.operand(1) just pass one parameter call.

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java (hunk 1 above)
Comment (vaibhavjain-dm) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:875]: Create a constant for literals. change all the places.

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 3]
```diff
@@ -885,6 +910,53 @@ private void unparseRegexpExtractAll(SqlWriter writer, SqlCall call,
     writer.endFunCall(regexpExtractAllFrame);
   }
 
+  private void unparseCot(SqlWriter writer, SqlCall call, int leftPrec, int rightPrec) {
+    final SqlWriter.Frame cotFrame = writer.startFunCall("1/tan");
```
Comment (vaibhavjain-dm) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:914]: 1/tan are two operands and one operator (/) so create SqlCall for 1 and tan and use operator divide.

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 4]
```diff
@@ -885,6 +910,53 @@ private void unparseRegexpExtractAll(SqlWriter writer, SqlCall call,
...
+      operandCall = SqlLiteral.createCharString(
+              unquoteStringLiteral(newOperand), SqlParserPos.ZERO);
+    }
+    final SqlWriter.Frame octetFrame = writer.startFunCall("OCTET_LENGTH");
+    operandCall.unparse(writer, leftPrec, rightPrec);
+    writer.endFunCall(octetFrame);
+  }
+
+  private void unparseInt2shFunctions(SqlWriter writer, SqlCall call,
+                                      String s, int leftPrec, int rightPrec) {
```
Comment (vaibhavjain-dm) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:939]: give proper name to variables.

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 5]
```diff
@@ -885,6 +910,53 @@ private void unparseRegexpExtractAll(SqlWriter writer, SqlCall call,
...
+    writer.print("(");
+    unparseBitwiseAnd(writer, call.operand(0), call.operand(2), leftPrec, rightPrec);
+    writer.sep(") " + s);
+    call.operand(1).unparse(writer, leftPrec, rightPrec);
+  }
+
+  private void unparseBiwiseFunctions(SqlWriter writer, SqlCall call,
+                                      String s, int leftPrec, int rightPrec) {
```
Comment (vaibhavjain-dm) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:947]: Give proper name to variables.

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 6]
```diff
@@ -885,6 +910,53 @@ private void unparseRegexpExtractAll(SqlWriter writer, SqlCall call,
...
+    call.operand(0).unparse(writer, leftPrec, rightPrec);
+    writer.sep(s);
+    call.operand(1).unparse(writer, leftPrec, rightPrec);
+  }
+
+  private void unparseBitwiseAnd(SqlWriter writer, SqlNode operand1,
+                                 SqlNode operand2, int leftPrec, int rightPrec) {
```
Comment (vaibhavjain-dm) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:954]: Give proper name to variables.

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 7]
```diff
@@ -864,6 +865,30 @@ private void unparseOtherFunction(SqlWriter writer, SqlCall call, int leftPrec,
...
+      unparseBiwiseFunctions(writer, call, "|", leftPrec, rightPrec);
+      break;
+    case "BITWISE_XOR":
+      unparseBiwiseFunctions(writer, call, "^", leftPrec, rightPrec);
+      break;
+    case "INT2SHR":
+      unparseInt2shFunctions(writer, call, ">>", leftPrec, rightPrec);
+      break;
+    case "INT2SHL":
+      unparseInt2shFunctions(writer, call, "<<", leftPrec, rightPrec);
```
Comment (vaibhavjain-dm) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:884]: rename unparseInt2shFunctions to unparseIntShiftOperation and revisit all the places.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 8]
```diff
@@ -3156,6 +3156,39 @@ public static Object regexpMatchCount(Object str, Object regex, Object startPos,
     }
     return count;
   }
+
+  public static Object cotFunction(Double value) {
+    return 1 / Math.tan(value);
+  }
+
+  public static Object bitwiseAnd(Integer value1, Integer value2) {
```
Comment (vaibhavjain-dm) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:3164]: value1 can be re-write as firstOperand same as value2.
Revisit all the variable declarations.

**Review by deepakDatametica — APPROVED** [review]

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java [hunk 9]
```diff
@@ -6780,6 +6780,99 @@ public void testCaseExprForEEE() {
     assertThat(toSql(root, DatabaseProduct.SNOWFLAKE.getDialect()), isLinux(expectedSF));
   }
 
+  @Test public void octetLength() {
```
Comment (vaibhavjain-dm) [inline core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:6783]: Rename test as testOctetLength

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java [hunk 10]
```diff
@@ -6780,6 +6780,99 @@ public void testCaseExprForEEE() {
...
+            builder.field("ENAME"));
+    final RelNode root = relBuilder().scan("EMP").filter(condition).build();
+
+    final String expectedBQ = "SELECT *\n"
+            + "FROM scott.EMP\n"
+            + "WHERE OCTET_LENGTH(ENAME)";
+    assertThat(toSql(root, DatabaseProduct.BIG_QUERY.getDialect()), isLinux(expectedBQ));
+  }
+
+  @Test public void octetLengthWithLiteral() {
```
Comment (vaibhavjain-dm) [inline core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:6795]: Rename test as testOctetLengthWithLiteral

File: core/src/test/java/org/apache/calcite/test/SqlFunctionsTest.java [hunk 11]
```diff
@@ -1164,6 +1171,38 @@ private void thereAndBack(byte[] bytes) {
     assertThat(SqlFunctions.monthsBetween("2019-05-26", "2020-04-20"), is(-10.806451613));
   }
 
+  @Test public void testCotFunction() {
```
Comment (vaibhavjain-dm) [inline core/src/test/java/org/apache/calcite/test/SqlFunctionsTest.java:1174]: Rename it as CotFunctionTest

Comment (rohitgaurDatametica) [reply core/src/test/java/org/apache/calcite/test/SqlFunctionsTest.java:1174]: done

**Review by vaibhavjain-dm — APPROVED** [review]

**Pull request merged**